### DISCUSSION TITLE
ResolutionContext.Items InvalidOperationException

### DISCUSSION THREAD

**Discussion by p-m-j** [LuckyPennySoftware/AutoMapper#4160] [NA]
Apologies as this is a straight duplicate of #3206 which appeared to reach consensus that something is not quite right however nothing has changed to rectify that. 

Whenever one makes use of ResolutionContext.Items there is a chance that the getter could throw InvalidOperationException depending on how Mapper.Map was called (with or without options action).

If there was a property on ResolutionContext that indicated that the instance was the default instance there would be a nice way to handle this e.g. in a IValueResolver so that Mapper.Map could be called both with and without the options action (without having to catch the exception and carry on).

### Source/destination types

```csharp
public class Source
{
    public int Id { get; set; }
    public string Name { get; set; }
}

public class Destination
{
    public int Id { get; set; }
    public string Name { get; set; }
}

```

### Mapping configuration

```csharp
class MaybeOverridesNameResolver : IValueResolver<Source, Destination, string>
{
    public string Resolve(Source source, Destination destination, string destMember, ResolutionContext context)
    {
        // context.Items getter may throw invalid operation
        if (context.Items.TryGetValue("name", out var overrideName))
        {
            return overrideName as string;
        }

        return source.Name;
    }
}

void Main()
{    
    var config = new MapperConfiguration(cfg =>
    {
        cfg.CreateMap<Source, Destination>()
            .ForMember(x => x.Name, dest => dest.MapFrom<MaybeOverridesNameResolver>());
    });

    var mapper = config.CreateMapper();
    var source = new Source { Id = 42, Name = "Foo" };

    // No problemo
    var a = mapper.Map<Destination>(source, _ => {} );

    // No problemo
    var b = mapper.Map<Destination>(source, ctx => ctx.Items["name"] = "Bar");
    
    // Throws as ResolutionContext is DefaultContext
    var c = mapper.Map<Destination>(source);
}

```

### Version: 
9+

**Comment by lbargaoanu** [comment]
https://github.com/AutoMapper/AutoMapper/discussions/4125

**Comment by p-m-j** [comment]
> I'd rather not add a flag for the sole purpose of telling you there is no context.

Could DefaultContext not just leave the property null and skip the invalid op?
No extra props but folk can null check

**Reply by lbargaoanu** [reply]
No, because people will ask why there's no context.

**Reply by p-m-j** [reply]
Is that worse than the current state where folk ask why it throws? :D
Nice little FAQ section on the docs site to cover it and you can close any issues like mine here with a swift RTFM

**Reply by lbargaoanu** [reply]
_Nothing_ is gonna stop people opening trivial issues :)
